tests/test_widgets.py

```diff
@@ -33,6 +51,74 @@ def test_after_translation_deactivate_all(self):
             '<select name="_0"><option value="" selected>---------</option>', rendered
         )
 
+    @override_settings(USE_I18N=True)
+    def test_value_from_datadict(self):
+        widget = PhoneNumberPrefixWidget()
+        tests = [
+            (
+                {
+                    "field_0": phonenumbers.region_code_for_number(number),
+                    "field_1": str(number.national_number),
+                },
+                number,
+            )
+            for number in EXAMPLE_NUMBERS
+        ]
+
+        for data, expected in tests:
+            with self.subTest(data):
+                self.assertEqual(
+                    widget.value_from_datadict(data, {}, "field"), expected
+                )
+
+    def test_value_from_datadict_invalid(self):
+        widget = PhoneNumberPrefixWidget()
+        data = {
+            "field_0": "US",
+            "field_1": "123",
+        }
+        value = widget.value_from_datadict(data, {}, "field")
+        self.assertEqual(value.as_e164, "+1123")
+        self.assertTrue(hasattr(value, "_region"))
+
+    def test_value_from_datadict_empty(self):
+        widget = PhoneNumberPrefixWidget()
+        data = {
+            "field_0": "US",
+            "field_1": "",
+        }
+        value = widget.value_from_datadict(data, {}, "field")
+        self.assertEqual(value.raw_input, "")
+        self.assertTrue(hasattr(value, "_region"))
+
+    def test_decompress_valid_str(self):
+        widget = PhoneNumberPrefixWidget()
+
+        tests = [
+            (
+                number.as_e164,
+                [
+                    phonenumbers.region_code_for_number(numobj=number),
+                    str(number.national_number),
+                ],
+            )
+            for number in EXAMPLE_NUMBERS
+        ]
+        for value, expected in tests:
+            with self.subTest(value):
+                self.assertEqual(widget.decompress(value=value), expected)
+
+    @override_settings(PHONENUMBER_DEFAULT_REGION="CA")
+    def test_decompress_invalid_str(self):
+        widget = PhoneNumberPrefixWidget()
+
+        valid_number = EXAMPLE_NUMBERS[0]
+        invalid_str = f"{valid_number.as_e164}1"
+
+        invalid_number = PhoneNumber.from_string(phone_number=invalid_str)
+        self.assertFalse(invalid_number.is_valid())
+        self.assertEqual(widget.decompress(invalid_number), ["CA", invalid_str])
```

Comment:
The value given to `decompress` [can be assumed to be valid](https://docs.djangoproject.com/en/4.0/ref/forms/widgets/#django.forms.MultiWidget.decompress), why is this test needed?

It may allow simplifying the `decompress` method.

Reply:
Thanks for the explanation, again makes total sense (and shows serious testing has gone into this PR :100:).

I wonder if we should update the test or add another one based on an actual view to better illustrate the use case. My concern is that the current test calls directly into `decompress` and it’s not obvious to developers how `decompress` would get called in that way from user interactions.

A docstring would be an alternative, although exercising the code is a better option. I’m thinking to instantiate the form with invalid user input (inconsistent prefix-number), verify it’s reported as invalid, and verify that the form rendered with errors keeps the region, the prefix and the invalid value the user supplied.
Happy to contribute the test if that’s helpful? What do you think?

Reply:
I wouldn't test it too much and on too high level because developers have some flexibility and they don't need to use `PhoneNumberPrefixWidget` straightforwardly on the one hand and we might end up testing Django internals on the other hand. That's why I focused myself on testing those parts which are overridden in the child class.

One level above a widget we have a form field which is a part of model from

Reply:
> It goes like this:
> 
>     1. user selects correct region and gives correct number but not in contexts of selected region
> 
>     2. `PhoneNumberPrefixWidget.value_from_datadict` is called and we get a `PhoneNumber` instance with invalid value
> 
>     3. `formfields.PhoneNumberField` is instantiated and we get a form error
> 
>     4. to render e.g. admin page with error message and request to correct given input `PhoneNumberPrefixWidget.decompress` is called with invalid `PhoneNumber` instance

I spent some time today trying to reproduce these STR and failed. Here’s my test project 
[test.tar.gz](https://github.com/stefanfoulis/django-phonenumber-field/files/8398566/test.tar.gz).

Here’s the output of pip-list:
```
Package                  Version                        Editable project location
------------------------ ------------------------------ -------------------------------------------
asgiref                  3.5.0
Babel                    2.9.1
black                    22.3.0
click                    8.1.2
Django                   4.0.3
django-phonenumber-field 6.1.1.dev12+g2a9a1d9.d20220401 /home/freitafr/dev/django-phonenumber-field
flake8                   4.0.1
isort                    5.10.1
mccabe                   0.6.1
mypy-extensions          0.4.3
pathspec                 0.9.0
phonenumberslite         8.12.46
pip                      22.0.4
platformdirs             2.5.1
pycodestyle              2.8.0
pyflakes                 2.4.0
pytz                     2022.1
setuptools               58.1.0
sqlparse                 0.4.2
tomli                    2.0.1
```

Here’s the patch I applied locally expecting a failure:
```diff
diff --git a/phonenumber_field/widgets.py b/phonenumber_field/widgets.py
index cd76de1..5dc8038 100644
--- a/phonenumber_field/widgets.py
+++ b/phonenumber_field/widgets.py
@@ -75,9 +75,9 @@ class PhoneNumberPrefixWidget(MultiWidget):
             value = to_python(value, region=region)
 
         if isinstance(value, PhoneNumber):
-            if not value.is_valid():
-                region_code = getattr(value, "_region", region)
-                return [region_code, value.raw_input]
+            # if not value.is_valid():
+            #     region_code = getattr(value, "_region", region)
+            #     return [region_code, value.raw_input]
             region_code = region_code_for_number(value)
             national_number = national_significant_number(value)
             return [region_code, national_number]
@@ -90,8 +90,6 @@ class PhoneNumberPrefixWidget(MultiWidget):
         if isinstance(number, str):
             number = PhoneNumber()
             number.raw_input = national_number
-        if not number.is_valid():
-            setattr(number, "_region", region_code)
         return number
 ```

Steps I took:
1. Create a user `./manage.py createsuperuser` 
2. Login as that user
3. http://127.0.0.1:8000/admin/app/mymodel/add/
4. Choose France as the country and enter invalid national number: `8876543210`

---


FWIW, the following test also passes with the patch:
```python
    @override_settings(PHONENUMBER_DEFAULT_REGION="CA")
    def test_decompress_invalid_str(self):
        class TestForm(forms.Form):
            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)

        invalid_national_number = "8876543210"
        form = TestForm(data={"phone_0": "FR", "phone_1": invalid_national_number})
        self.assertIs(form.is_valid(), False)
        rendered_form = str(form)
        # Retain selected region.
        self.assertInHTML(
            """
            <ul class="errorlist">
            <li>
                Enter a valid phone number (e.g. (506) 234-5678) or a number
                with an international call prefix.
            </li>
            </ul>
            """,
            rendered_form,
        )
        self.assertInHTML(
            '<option value="FR" selected>France +33</option>',
            rendered_form,
            count=1,
        )
        self.assertInHTML(
            f'<input type="text" name="phone_1" value="{invalid_national_number}" required id="id_phone_1">',
            rendered_form,
            count=1,
        )
```

---

I am still unclear on when decompress is passed invalid data?

Reply:
I just used your project and latest state of this PR (commmit 2a9a1d9). I added in [widgets.py](phonenumber_field/widgets.py) a new raising exception line after line 78 like this:

```python
        if isinstance(value, PhoneNumber):
            if not value.is_valid():
                raise ValueError(value)
                region_code = getattr(value, "_region", region)
                return [region_code, value.raw_input]
```
instead of this
```python
        if isinstance(value, PhoneNumber):
            if not value.is_valid():
                region_code = getattr(value, "_region", region)
                return [region_code, value.raw_input]
```

Then I navigated to `/admin/app/mymodel/add/` URL and selected `France +33` and I hit `SAVE`. Please note empty national number and voilà! I've got pretty nasty exception as expected.

This also works for let say `France +33` and `000000`.

Reply:
And your patch rolls back changes which help for #461.

Reply:
What I am looking for is noticeable issues from a user perspective, to write tests that map reasonably closely to usage instead of being coupled to the implementation. That’s my main concern with the `test_decompress`. That’s why I removed some bits of code and tried using the app, expecting a failure. I was not suggesting to apply my patch to this PR, sorry for the misunderstanding.
Adding an exception to show the code path is hit does not help in building a test case that’s representative of the expected behavior.

With the patch above, the invalid form is displayed correctly, with the selected country code and the national number. That’s why I was asking for the kind of failures the STR were addressing.

> Then I navigated to /admin/app/mymodel/add/ URL and selected France +33 and I hit SAVE. Please note empty national number [...]

That’s the kind of info I was looking for! In trying out more today, I found a couple issues captured in the following tests (only two tests show incorrect behavior from the system, the others are provided because I think they should be added to the test suite):
```diff
diff --git a/tests/test_widgets.py b/tests/test_widgets.py
index d632e71..fa1804e 100644
--- a/tests/test_widgets.py
+++ b/tests/test_widgets.py
@@ -1,8 +1,10 @@
 import phonenumbers
+from django import forms
 from django.test import SimpleTestCase, override_settings
 from django.utils import translation
 
 from phonenumber_field.phonenumber import PhoneNumber
+from phonenumber_field import formfields
 from phonenumber_field.widgets import (
     PhoneNumberInternationalFallbackWidget,
     PhoneNumberPrefixWidget,
@@ -107,16 +109,103 @@ class PhoneNumberPrefixWidgetTest(SimpleTestCase):
             with self.subTest(value):
                 self.assertEqual(widget.decompress(value), expected)
 
-    @override_settings(PHONENUMBER_DEFAULT_REGION="CA")
-    def test_decompress_invalid_str(self):
-        widget = PhoneNumberPrefixWidget()
+    def test_empty_region(self):
+        class TestForm(forms.Form):
+            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)
+
+        invalid_national_number = "8876543210"
+        form = TestForm(data={"phone_0": "", "phone_1": invalid_national_number})
+        self.assertIs(form.is_valid(), False)
+        rendered_form = str(form)
+        self.assertInHTML(
+            """
+            <ul class="errorlist"><li>Enter a valid phone number (e.g. +12125552368).</li></ul>
+            """,
+            rendered_form,
+        )
+        # Keeps national number input.
+        self.assertInHTML(
+            f'<input type="text" name="phone_1" value="{invalid_national_number}" required id="id_phone_1">',
+            rendered_form,
+            count=1,
+        )
+        self.assertInHTML(
+            '<option value="" selected>---------</option>',
+            rendered_form,
+            count=1,
+        )
+
+    def test_empty_national_number(self):
+        class TestForm(forms.Form):
+            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)
+
+        form = TestForm(data={"phone_0": "CA", "phone_1": ""})
+        self.assertIs(form.is_valid(), False)
+        rendered_form = str(form)
+        self.assertInHTML(
+            """
+            <ul class="errorlist"><li>Enter a valid phone number (e.g. +12125552368).</li></ul>
+            """,
+            rendered_form,
+        )
+        self.assertInHTML(
+            f'<input type="text" name="phone_1" required id="id_phone_1">',
+            rendered_form,
+            count=1,
+        )
+        self.assertInHTML(
+            '<option value="CA" selected>Canada +1</option>',
+            rendered_form,
+            count=1,
+        )
 
-        valid_number = EXAMPLE_NUMBERS[0]
-        invalid_str = f"{valid_number.as_e164}1"
+    def test_no_region(self):
+        class TestForm(forms.Form):
+            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)
+
+        form = TestForm(data={"phone_1": "654321"})
+        self.assertIs(form.is_valid(), False)
+        rendered_form = str(form)
+        self.assertInHTML(
+            """
+            <ul class="errorlist"><li>Enter a valid phone number (e.g. +12125552368).</li></ul>
+            """,
+            rendered_form,
+        )
+        self.assertInHTML(
+            f'<input type="text" name="phone_1" value="654321" required id="id_phone_1">',
+            rendered_form,
+            count=1,
+        )
+        self.assertInHTML(
+            '<option value="" selected>---------</option>',
+            rendered_form,
+            count=1,
+        )
 
-        invalid_number = PhoneNumber.from_string(invalid_str)
-        self.assertFalse(invalid_number.is_valid())
-        self.assertEqual(widget.decompress(invalid_number), ["CA", invalid_str])
+    def test_no_national_number(self):
+        class TestForm(forms.Form):
+            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)
+
+        form = TestForm(data={"phone_0": "CA"})
+        self.assertIs(form.is_valid(), False)
+        rendered_form = str(form)
+        self.assertInHTML(
+            """
+            <ul class="errorlist"><li>Enter a valid phone number (e.g. +12125552368).</li></ul>
+            """,
+            rendered_form,
+        )
+        self.assertInHTML(
+            f'<input type="text" name="phone_1" value="" required id="id_phone_1">',
+            rendered_form,
+            count=1,
+        )
+        self.assertInHTML(
+            '<option value="CA" selected>Canada +1</option>',
+            rendered_form,
+            count=1,
+        )
 
 
 class PhoneNumberInternationalFallbackWidgetTest(SimpleTestCase):
```

```
======================================================================
ERROR: test_no_national_number (tests.test_widgets.PhoneNumberPrefixWidgetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/freitafr/dev/django-phonenumber-field/tests/test_widgets.py", line 191, in test_no_national_number
    self.assertIs(form.is_valid(), False)
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/forms.py", line 205, in is_valid
    return self.is_bound and not self.errors
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/forms.py", line 200, in errors
    self.full_clean()
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/forms.py", line 433, in full_clean
    self._clean_fields()
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/forms.py", line 440, in _clean_fields
    value = bf.initial if field.disabled else bf.data
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/boundfield.py", line 127, in data
    return self.form._widget_data_value(self.field.widget, self.html_name)
  File "/home/freitafr/dev/django-phonenumber-field/.tox/py310/lib/python3.10/site-packages/django/forms/forms.py", line 224, in _widget_data_value
    return widget.value_from_datadict(self.data, self.files, html_name)
  File "/home/freitafr/dev/django-phonenumber-field/phonenumber_field/widgets.py", line 93, in value_from_datadict
    if not number.is_valid():
AttributeError: 'NoneType' object has no attribute 'is_valid'

======================================================================
FAIL: test_empty_national_number (tests.test_widgets.PhoneNumberPrefixWidgetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/freitafr/dev/django-phonenumber-field/tests/test_widgets.py", line 143, in test_empty_national_number
    self.assertIs(form.is_valid(), False)
AssertionError: True is not False
```

---

> And your patch rolls back changes which help for https://github.com/stefanfoulis/django-phonenumber-field/issues/461.

Yes, because I was also trying to catch the incorrect behavior without the private `_region` attr. The current tests rely on an implementation detail, which is not as helpful as verifying the expected behavior. The presence of the attribute does not indicate that it is used, nor its purpose.
Here’s an example test verifying the behavior.
```python
    def test_keeps_region_with_invalid_national_number(self):
        class TestForm(forms.Form):
            phone = formfields.PhoneNumberField(widget=PhoneNumberPrefixWidget)

        form = TestForm(data={"phone_0": "CA", "phone_1": "0000"})
        self.assertIs(form.is_valid(), False)
        rendered_form = str(form)
        self.assertInHTML(
            """
            <ul class="errorlist"><li>Enter a valid phone number (e.g. +12125552368).</li></ul>
            """,
            rendered_form,
        )
        self.assertInHTML(
            f'<input type="text" name="phone_1" value="0000" required id="id_phone_1">',
            rendered_form,
            count=1,
        )
        self.assertInHTML(
            '<option value="CA" selected>Canada +1</option>',
            rendered_form,
            count=1,
        )
```

Reply:
Please find your patch slightly updated as a part of my recent commits. I really appreciate those examples you gave because now it's clearer that you are aiming for functional tests over unit tests. I let myself to improve few lines here and there but all kudos goes to you.

I also discovered that in very edge case scenario which is rather not probably `None` value can be returned in `value_from_datadict` causing the exception you referred to. This is fixed in a9ed219. There is one more issue with validation where invalid `PhoneNumber` instance with `raw_input` property set to empty string is evaluated to `False` in `if` clause in: https://github.com/stefanfoulis/django-phonenumber-field/blob/022dccc3caffe3aba0ea8388fb3d4a2201f2f0fb/phonenumber_field/validators.py#L7-L12
This is also fixed in c05200a.

I hope this exhaust all uncertainties in this thread. I'm really looking forward to your feedback on recent improvements. I also hope this makes this PR complete.